receiver/snmpreceiver/config_helper.go

```diff
@@ -0,0 +1,297 @@
+	}
+
+	return dataPointAttributes
+}
+
+// getDataPointAttributes gets attributes for this metric's datapoint based on the previously
+// gathered attributes.
+// Keys will be determined from the related attribute config and values will come a few
+// different places.
+// Enum attribute value - comes from the metric config's attribute data
+// Indexed prefix attribute value - comes from the current SNMP data's index and the attribute
+// config's prefix value
+// Indexed OID attribute value - comes from the previously collected indexed attribute data
+// using the current index and attribute config to access the correct value
+func (h configHelper) getIndexedDataPointAttributes(
+	oid string,
+	parentOID string,
+	indexedAttributeValues map[indexedAttributeKey]string,
+) (map[string]string, error) {
+	datapointAttributes := map[string]string{}
+	indexString := strings.TrimPrefix(oid, parentOID)
+
+	for _, attribute := range h.metricAttributesByOID[parentOID] {
+		attributeCfg := h.cfg.Attributes[attribute.Name]
+		attributeKey := attribute.Name
+		if attributeCfg.Value != "" {
+			attributeKey = attributeCfg.Value
+		}
+		attributeValue := attribute.Value
+		if attributeCfg.IndexedValuePrefix != "" {
+			attributeValue = attributeCfg.IndexedValuePrefix + indexString
+		} else if attributeCfg.OID != "" {
+			attrKey := indexedAttributeKey{
+				parentOID: attributeCfg.OID,
+				oidIndex:  indexString,
+			}
+			attributeValue = indexedAttributeValues[attrKey]
+			if attributeValue == "" {
+				return nil, errors.New(errMsgAttributeEmptyValue)
+			}
+		}
+		datapointAttributes[attributeKey] = attributeValue
+	}
+
+	return datapointAttributes, nil
+}
+
+// getResourceAttributes creates a map of key/values for all related resource attributes. Keys
+// will come directly from the metric config's resource attribute values. Values will come
+// from the related attribute config's prefix value plus the index OR the previously collected
+// resource attribute indexed data.
+func (h configHelper) getResourceAttributes(
+	oid string,
+	parentOID string,
+	indexedResourceAttributeValues map[indexedAttributeKey]string,
+) (map[string]string, error) {
+	resourceAttributes := map[string]string{}
+	indexString := strings.TrimPrefix(oid, parentOID)
+
+	for _, attributeName := range h.resourceAttributesByOID[parentOID] {
+		resourceAttributeCfg := h.cfg.ResourceAttributes[attributeName]
+
+		var attributeValue string
+		if resourceAttributeCfg.IndexedValuePrefix != "" {
+			attributeValue = resourceAttributeCfg.IndexedValuePrefix + indexString
+		} else if resourceAttributeCfg.OID != "" {
+			attrKey := indexedAttributeKey{
+				parentOID: resourceAttributeCfg.OID,
+				oidIndex:  indexString,
+			}
+			attributeValue = indexedResourceAttributeValues[attrKey]
+			if attributeValue == "" {
+				return nil, errors.New(errMsgResourceAttributeEmptyValue)
+			}
+		}
+		resourceAttributes[attributeName] = attributeValue
+	}
+
+	return resourceAttributes, nil
+}
```

Comment:
These two wound up doing a lot of the higher-level processing of data. I'm not sure exactly how best to untangle that, but a few thoughts:

---

You have this pattern in both functions:
```go
	for _, attributeName := range h.resourceAttributesByOID[parentOID] {
		resourceAttributeCfg := h.cfg.ResourceAttributes[attributeName]
```

Pull the loop back out into the scraper so most of the processing can occur there. Use this struct to pull up what you need from it.

```go
	// in scraper.go
	for _, attributeName := range cfgHelper.getResourceAttributeColumnOIDs(parentOID) {
		resourceAttributeCfg := cfgHelper.getResourceAttributes(attributeName)
```

---

Isolating the indexString is related to processing of data, so parse that in the scraper.

---

Don't pass in indexedResourceAttributeValues either, since this is data retrieved from the client. 

---

It seems like there is probably a way to split the inner logic so the config is handling only very simple config-related things, and the scraper is asking the config for only what it needs. That said, I see that the boundary here is a but murkier, so don't worry too much. Err on the side of having the scraper process the data though, rather than the config helper.

Reply:
@djaglowski i appreciate all that you are saying here. in my opinion, these functions basically should exist in this helper almost entirely or they should exist in the scraper almost entirely.

if the config helper is going to be a completely "pure" class, then these functions shouldn't be here and only the most basic getters could be accessed through the configHelper rather than using the config directly (just as you suggested). i don't think there is much to tease apart other than what you already suggested.  at the same time...it feels kind of weird that there is a getScalarAttributes function which could completely exist within this helper (because it only uses the config given a passed in OID), but its cousins (getting indexed attributes or resource attributes) would live elsewhere.

or you let this helper be a little more loosey goosey and be able to take in some data format defined outside of it in order to have all 3 functions located together. we could also have a 3rd helper wraps the config helper and then there is less issue with keeping it so "clean"...but then this sounds like overcomplicating a problem that isn't too large.

so in the end i'll end up doing exactly what you suggested as you're the gatekeeper here even though i don't love that these 3 functions won't all live in the same place. i "get it", but i don't fully like it haha. i just would like to have this finished sooner than later.

Reply:
> it feels kind of weird that there is a getScalarAttributes function which could completely exist within this helper (because it only uses the config given a passed in OID), but its cousins (getting indexed attributes or resource attributes) would live elsewhere

I agree this isn't ideal. I'm ok with moving getScalerAttributes back out. The helper class will still be useful for clarifying the difference between user configuration and data, which is what I really care about here.